server/src/main/java/org/elasticsearch/injection/PlanInterpreter.java

```diff
@@ -0,0 +1,105 @@
+ * are supposed to be detected during execution; they should be detected during planning and validation.
+ * All exceptions thrown during execution are supposed to be caused by user-supplied code.
+ *
+ * <p>
+ * <strong>Execution model</strong>:
+ * The state of the injector during injection comprises a map from classes to objects.
+ * Before any steps execute, the map is pre-populated by object instances added via
+ * {@link Injector#addInstance(Object)}  Injector.addInstance},
+ * and then the steps begin to execute, reading and writing from this map.
+ * Some steps create objects and add them to this map; others manipulate the map itself.
+ */
+final class PlanInterpreter {
+    private static final Logger logger = LogManager.getLogger(PlanInterpreter.class);
+    private final Map<Class<?>, Object> instances = new LinkedHashMap<>();
+
+    PlanInterpreter(Map<Class<?>, Object> existingInstances) {
+        existingInstances.forEach(this::addInstance);
+    }
+
+    /**
+     * Main entry point. Contains the implementation logic for each {@link InjectionStep}.
+     */
+    void executePlan(List<InjectionStep> plan) {
+        int numConstructorCalls = 0;
+        for (InjectionStep step : plan) {
+            if (step instanceof InstantiateStep i) {
+                MethodHandleSpec spec = i.spec();
+                logger.trace("Instantiating {}", spec.requestedType().getSimpleName());
+                addInstance(spec.requestedType(), instantiate(spec));
+                ++numConstructorCalls;
+            } else {
+                // TODO: switch patterns would make this unnecessary
+                assert false : "Unexpected step type: " + step.getClass().getSimpleName();
+                throw new IllegalStateException("Unexpected step type: " + step.getClass().getSimpleName());
+            }
+        }
+        logger.debug("Instantiated {} objects", numConstructorCalls);
+    }
+
+    /**
+     * @return the list element corresponding to instances.get(type).get(0),
+     * assuming that instances.get(type) has exactly one element.
+     * @throws IllegalStateException if instances.get(type) does not have exactly one element
+     */
+    public <T> T theInstanceOf(Class<T> type) {
+        Object instance = instances.get(type);
+        if (instance == null) {
+            throw new IllegalStateException("No object of type " + type.getSimpleName());
+        }
+        return type.cast(instance);
+    }
+
+    private void addInstance(Class<?> requestedType, Object instance) {
+        Object old = instances.put(requestedType, instance);
+        if (old != null) {
+            throw new IllegalStateException("Multiple objects for " + requestedType);
+        }
+    }
+
+    /**
+     * @throws IllegalStateException if the <code>MethodHandle</code> throws.
+     */
+    @SuppressForbidden(reason = "Can't call invokeExact because we don't know the method argument types statically")
```

Comment:
Can you put a bit more information on this in the doc/comments? It feels like this is a key point that needs to be understood.

Reply:
Sure, do you have a suggestion? I'm not sure what aspect is unclear.

Reply:
Why don't we know the method types statically? What do we know? What conversions need to take place as part of the invoke call such that `invokeExact` is insufficient?

Reply:
Ah I see. I can add that.